pkg/security/probe/load_controller.go

```diff
@@ -0,0 +1,154 @@
+	"github.com/hashicorp/golang-lru/simplelru"
+
+	"github.com/DataDog/datadog-agent/pkg/util/log"
+)
+
+type eventCounterLRUKey struct {
+	Pid   uint32
+	Event EventType
+}
+
+// LoadController is used to monitor and control the pressure put on the host
+type LoadController struct {
+	sync.RWMutex
+	probe        *Probe
+	total        int64
+	counters     *simplelru.LRU
+	statsdClient *statsd.Client
+
+	EventsCountThreshold int64
+	DiscarderTimeout     time.Duration
+	ControllerPeriod     time.Duration
+}
+
+// NewLoadController instantiates a new load controller
+func NewLoadController(probe *Probe, statsdClient *statsd.Client) (*LoadController, error) {
+	lru, err := simplelru.NewLRU(probe.config.PIDCacheSize, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	lc := &LoadController{
+		probe:                probe,
+		counters:             lru,
+		statsdClient:         statsdClient,
+		EventsCountThreshold: probe.config.LoadControllerEventsCountThreshold,
+		DiscarderTimeout:     probe.config.LoadControllerDiscarderTimeout,
+		ControllerPeriod:     probe.config.LoadControllerControlPeriod,
+	}
+	return lc, nil
+}
+
+// Count increments the event counter of the provided event type and pid
+func (lc *LoadController) Count(eventType EventType, pid uint32) {
+	lc.Lock()
+	defer lc.Unlock()
+
+	entry, ok := lc.counters.Get(eventCounterLRUKey{Pid: pid, Event: eventType})
+	if ok {
+		counter := entry.(*uint64)
+		atomic.AddUint64(counter, 1)
+	} else {
+		count := uint64(1)
+		lc.counters.Add(eventCounterLRUKey{Pid: pid, Event: eventType}, &count)
+	}
+	newTotal := atomic.AddInt64(&lc.total, 1)
+
+	if newTotal >= lc.EventsCountThreshold {
+		lc.discardNoisiestProcess()
+	}
+}
+
+// discardNoisiestProcess determines the noisiest process and event_type tuple and pushes a temporary discarder
+func (lc *LoadController) discardNoisiestProcess() {
+	// iterate over the LRU map to retrieve the noisiest process & event_type tuple
+	var maxKey eventCounterLRUKey
+	var maxCount *uint64
+	for _, key := range lc.counters.Keys() {
+		entry, ok := lc.counters.Peek(key)
+		if !ok || entry == nil {
+			continue
+		}
+		tmpCount := entry.(*uint64)
+		tmpKey := key.(eventCounterLRUKey)
+
+		// update max if necessary
+		if maxCount == nil || *maxCount < *tmpCount {
+			maxCount = tmpCount
+			maxKey = tmpKey
+		}
+	}
+	if maxCount == nil {
+		// LRU is empty nothing to discard
+		return
+	}
+
+	// push a temporary discarder on the noisiest process & event type tuple
+	log.Debugf("discarding %s events from pid %d for %s seconds", maxKey.Event, maxKey.Pid, lc.DiscarderTimeout)
+	if _, err := discardPIDWithTimeout(lc.probe, maxKey.Event, maxKey.Pid, lc.DiscarderTimeout); err != nil {
+		log.Debugf("couldn't insert temporary discarder: %v", err)
+		return
+	}
+
+	// update current total and remove biggest entry from cache
+	atomic.AddInt64(&lc.total, -int64(atomic.SwapUint64(maxCount, 0)))
+
+	// send load_controller.pids_discarder metric
+	tags := []string{
```

Comment:
Could we also send the PID, and/or the process name?

Reply:
Technically we could, unfortunately we decided not to because the cardinality of such a tag would explode and be uncontrollable 😢